### PR TITLE
WT-11468 Fix the ordering when resolving truncate prepare updates

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: src/include/txn_inline.h [hunk 1]
```diff
@@ -251,22 +302,18 @@ __wt_txn_op_delete_apply_prepare_state(WT_SESSION_IMPL *session, WT_REF *ref, bo
...
-                 */
-                (*updp)->prepare_state = prepare_state;
                 if (commit)
-                    (*updp)->durable_ts = txn->durable_timestamp;
+                    __txn_resolve_prepared_update(session, *updp);
+                else
+                    __txn_prepare_update(session, *updp);
             }
     }
     page_del = ref->page_del;
```
Comment (luke-pearson) [inline src/include/txn_inline.h:311]: Do we need a volatile cast here? This looks like READ_ONCE is needed to me.

Comment (quchenhao) [reply src/include/txn_inline.h:311]: No, I don't think it is required here. The ref is locked here and other thread can change anything in the ref.

File: src/include/txn_inline.h [hunk 2]
```diff
@@ -152,6 +194,27 @@ __txn_resolve_prepared_update(WT_SESSION_IMPL *session, WT_UPDATE *upd)
...
+__txn_resolve_prepared_page_deleted(WT_SESSION_IMPL *session, WT_PAGE_DELETED *page_del)
+{
+    WT_TXN *txn;
+
+    txn = session->txn;
+
+    /*
+     * The page deleted structure is only checked in tree walk. If it is prepared, we will
+     * instantiate the leaf page and check the keys on it. Therefore, we don't need to worry about
+     * reading the partial state and thus no need to lock the state.
```
Comment (luke-pearson) [inline src/include/txn_inline.h:211]: and thus no need -> and don't need

File: src/include/txn_inline.h [hunk 3]
```diff
@@ -152,6 +194,27 @@ __txn_resolve_prepared_update(WT_SESSION_IMPL *session, WT_UPDATE *upd)
     WT_PUBLISH(upd->prepare_state, WT_PREPARE_RESOLVED);
 }
 
+/*
+ * __txn_resolve_prepared_page_deleted --
+ *     Resolve a prepared page deleted structure.
+ */
+static inline void
+__txn_resolve_prepared_page_deleted(WT_SESSION_IMPL *session, WT_PAGE_DELETED *page_del)
```
Comment (luke-pearson) [inline src/include/txn_inline.h:202]: Why not a function like: 
```__txn_apply_prepare_state_page_del(WT_SESSION_IMPL *session, WT_PAGE_DELETED *page_del, bool commit) {
     WT_TXN *txn;

     txn = session->txn;

     if (commit) {
        page_del->timestamp  = txn->commit_timestamp;
        page_del->durable_timestamp = txn->durable_timestamp;
        WT_PUBLISH(page_del->prepare_state, WT_PREPARE_RESOLVED)
       } else {
          upd->durable_ts = WT_TS_NONE;
        WT_PUBLISH(upd->prepare_state, WT_PREPARE_INPROGRESS);
       upd->durable_ts = WT_TS_NONE;
        WT_PUBLISH(upd->prepare_state, WT_PREPARE_INPROGRESS);
      }
   }

```

Comment (quchenhao) [reply src/include/txn_inline.h:202]: I don't want to mix the prepare logic and commit logic in the same function. We do them separately in normal update so I also want to keep them separately here.

Comment (luke-pearson) [reply src/include/txn_inline.h:202]: I think it'll be cleaner in the calling code if we pass commit in, e.g:
```
   for (; *updp != NULL; ++updp)
       txn_apply_prepare_state_update(session, *updp, commit);
  
   txn_apply_prepare_state_page_del(session, ref->page_del, commit);
```

Comment (kommiharibabu) [reply src/include/txn_inline.h:202]: I agree with Luke. Change the existing update logic also into a single function. The suggested change is much cleaner than the existing one.

Comment (quchenhao) [reply src/include/txn_inline.h]: done

File: src/include/txn_inline.h [hunk 4]
```diff
@@ -251,23 +292,15 @@ __wt_txn_op_delete_apply_prepare_state(WT_SESSION_IMPL *session, WT_REF *ref, bo
...
     }
-    page_del = ref->page_del;
-    if (page_del != NULL) {
-        page_del->timestamp = ts;
-        if (commit)
-            page_del->durable_timestamp = txn->durable_timestamp;
-        WT_PUBLISH(page_del->prepare_state, prepare_state);
-    }
+
+    if (ref->page_del != NULL)
```
Comment (luke-pearson) [inline src/include/txn_inline.h:302]: You could move that into the function call, I don't mind either way. I also think we should do the same shared function for regular updates not just page_del structs.

Comment (quchenhao) [reply src/include/txn_inline.h:291]: done

File: src/include/txn_inline.h [hunk 5]
```diff
@@ -250,25 +285,13 @@ __wt_txn_op_delete_apply_prepare_state(WT_SESSION_IMPL *session, WT_REF *ref, bo
...
-        page_del->timestamp = ts;
-        if (commit)
-            page_del->durable_timestamp = txn->durable_timestamp;
-        WT_PUBLISH(page_del->prepare_state, prepare_state);
+            for (; *updp != NULL; ++updp)
+                __txn_apply_prepare_state_update(session, *updp, commit);
     }
 
+    if ((page_del = ref->page_del) != NULL)
+        __txn_apply_prepare_state_page_del(session, page_del, commit);
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/txn_inline.h:293]: It bothers me if this is null, why we need to apply the change. Anyway, I'll change if you insist.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/include/txn_inline.h:293]: Can you elaborate on why that bothers you if theres a good argument against it I'm okay with leaving it as is.

[the remark this answers is not in the document]
Comment (quchenhao) [reply src/include/txn_inline.h:293]: I feel the logic is clearer in the caller. Skipping the "work" if null is not technically the "work". Also if we do the null check in the function, we lose the certainty that we don't need a read once logic here if some other function calls the function.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/include/txn_inline.h:293]: I agree with that argument, happy to keep it as is.

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Comment by quchenhao** [comment]
evergreen merge

**Pull request merged**